facebook/rocksdb
# MultiGet parallel IO

Enhancement to MultiGet batching to read data blocks required for keys in a batch in parallel from disk. It uses Env::MultiRead() API to read multiple blocks and reduce latency.

Test Plan:
1. make check
2. make asan_check
3. make asan_crash

Perf results:
For a 100% cache workload, there is a minimal regression of ~1.5%. The followed command was used -
```
TEST_TMPDIR=/dev/shm/multiget numactl -C 10  ./db_bench -use_existing_db=true -benchmarks="readseq,multireadrandom" -write_buffer_size=16777216 -target_file_size_base=4194304 -max_bytes_for_level_base=16777216 -num=12000000 -reads=12000000 -duration=90 -threads=1 -compression_type=none -cache_size=4194304000 -batch_size=32 -disable_auto_compactions=true -bloom_bits=10 -cache_index_and_filter_blocks=true -pin_l0_filter_and_index_blocks_in_cache=true -multiread_batched=true -multiread_stride=4
```
```
readrandom   :       4.107 micros/op 243488 ops/sec;   26.9 MB/s (21913999 of 21913999 found)
multireadrandom (w/o this PR):       2.161 micros/op 462658 ops/sec; (41663968 of 41663968 found)
multireadrandom  (w this PR):       2.191 micros/op 456325 ops/sec; (41087968 of 41087968 found)
```


For a IO heavy workload (i.e nearly every data block is a miss), the following command was used. The results are given below.
```
TEST_TMPDIR=/data/rocksdb/multiget numactl -C 10  ./db_bench -use_existing_db=true -benchmarks="${benchmark}" -write_buffer_size=67108864 -target_file_size_base=67108864 -max_bytes_for_level_base=1073741824 -num=12000000 -reads=12000000 -duration=60 -threads=1 -compression_type=none -cache_size=41943040000 -batch_size=$batch_size -disable_auto_compactions=true -bloom_bits=10 -multiread_batched=true -multiread_stride=$stride -seed `od -A n -t u -N 4 /dev/urandom`
```
A stride of 16 ensured most keys in a batch were in same SST file, whereas stride of 4096 makes every key fall in a different file.

```
 		                                  Batch Size					
Stride	Benchmark	          2      4      8      16       32       64
 		                                  (micros/op)
16	Get	               278.752  277.303	277.861	278.785	277.751	275.792
	MultiGet	       279.423	274.534	273.095	271.037	271.616	269.6
	MultiGet w/ MultiRead  205.826	151.89	116.589	95.589	84.32	83.057
							
4096	Get	               283.642	281.951	277.928	276.1	278.004	269.064
	MultiGet	       281.612  281.253 276.227 276.462 272.94  270.359
	MultiGet w/ MultiRead	280.49	276.696	275.585	273.388	269.661	270.916
```
